pkg/apply/applier_builder.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2021 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package apply
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/cli-utils/pkg/apply/info"
+	"sigs.k8s.io/cli-utils/pkg/apply/poller"
+	"sigs.k8s.io/cli-utils/pkg/apply/prune"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ApplierBuilder struct {
+	// factory is only used to retrieve things that have not been provided explicitly.
+	factory                      util.Factory
```

Comment:
Do we need all these options? I'm not even sure which ones are mutually exclusive. I think some of it we might be able to clean up by simplifying things internally. I'm also uncertain if we should let users provide things like the RESTMapper, since we in practice have some requirements on the implementation that isn't obvious.

Reply:
I think it's good to allow users to inject dependencies directly. For those who don't need to do that, they can inject just the factory and nothing else. My main motivator is the desire to inject a custom `rest.Config`/`dynamic.Interface` with a dedicated rate limiter. Other parts of my program should use a separate rate limiter from the applier. I could do it by creating a custom factory, but that is quite unwieldy. It's an example use case, anyway.

> I'm also uncertain if we should let users provide things like the RESTMapper, since we in practice have some requirements on the implementation that isn't obvious.

If there is a requirement, ideally code should fail early if it's not met rather than silently not do what it is supposed to (current behavior). I'm going to fix that in https://github.com/kubernetes-sigs/cli-utils/issues/421 by using code from https://github.com/kubernetes/kubernetes/pull/105623. As part of that I'm going to change `meta.RESTMapper` to `meta.ResettableRESTMapper` here so the requirement is explicit.